### ISSUE TITLE
Receiving a failed to compile - Module not found: Can't resolve 'ws' in WebSocketTransport.js

### ISSUE THREAD

**Issue by zhg-kj** [BlakeStevenson/populartimes.js#1]
I ran npm install websocket and puppeteer. However, I still receive a cannot resolve ws error. Any idea on what I might have missed?

**Comment by BlakeStevenson** [comment]
Are you running it locally or on a host? Some hosts don’t allow puppeteer, since it requires a lot of system resources. Try following the instructions in the README file, run npm install populartimes.js then require it in your file.

**Comment by zhg-kj** [comment]
I am running it through ReactJS which uses my localhost(Puppeteer should work on React really well according to google). I have followed all the instructions in the README file. Here is a snippet of the code:
```
import React, { Component } from 'react';
import { Map, GoogleApiWrapper, InfoWindow, Marker } from 'google-maps-react';

const key = '';

let temp;
const populartimes = require('populartimes.js');
populartimes('ChIJKznozuF0hlQRbfbe7fJJ1rM').then(out => {temp=out.now.percent});
```

**Comment by BlakeStevenson** [comment]
From my cursory google search, Puppeteer doesn't work in the browser, and "using puppeteer in a react website" only results in how to automate app testing, not how to use it within the app itself. If you want to use puppeteer in the app itself, you would need to use a third-party service like https://www.browserless.io, here is a modified version of the script that uses browserless (you can just copy and paste it into the `index.js` file in `node_modules/populartimes.js` directory. You'll need to run `npm remove puppeteer` then `npm install puppeteer-core`. Keep in mind that without authentication, browserless only allows you to make 50 requests per day.

```
const jsdom = require("jsdom");
const { JSDOM } = jsdom;
const puppeteer = require('puppeteer-core');
//https://www.google.com/maps/place/?q=place_id:ChIJEVBPhRQtTIYR9Qn5LawiZIs

function getHtmlUrl(placeId) {
    return `https://www.google.com/maps/place/?q=place_id:${placeId}`;
}

function getDayName(index) {
    const days = {
        0: "sunday",
        1: "monday",
        2: "tuesday",
        3: "wednesday",
        4: "thursday",
        5: "friday",
        6: "saturday",
    }
    return days[index];
}

async function sendRequest(htmlUrl, params) {
    const browser = await puppeteer.connect({ 
      browserWSEndpoint: `wss://chrome.browserless.io` 
    });
    const page = await browser.newPage();
    await page.goto(htmlUrl, {waitUntil: 'networkidle2'});
    const data = await page.evaluate(() => document.querySelector('*').outerHTML);
    await browser.close();
    return data;
}

module.exports = async function getPopularTimes(placeId) {
    // get raw html
    const rawData = await sendRequest(getHtmlUrl(placeId));
    // parse html
    const body = new JSDOM(rawData);
    // get days of the week
    const days = body.window.document.getElementsByClassName("section-popular-times-graph");
    // loop through the days
    const out = {};
    let i = 0;
    for(let day of days) {
        // get hours in a day
        const hours = day.getElementsByClassName("section-popular-times-bar");
        // loop through the hours
        let hoursInDay = [];
        let j = 0;
        for(let hour of hours) {
            let hr = hour.getAttribute("aria-label");
            let parts = hr.split(" ");
            if(parts[0] === "Currently") {
                out.now = {
                    percent: parts[4],
                }
            } else {
                let percent = parts[0];
                let hour = parts[3];
                let meridiem = parts[4].replace(".", "");

                hoursInDay[j] = {percent, hour, meridiem};
                j++;
            }
        }
        out[getDayName(i)] = hoursInDay;
        i++;
    }
    return out;
}
```

**Comment by zhg-kj** [comment]
Still running into the same issue where ws does not resolve after doing that. It seems that I might have to create a Node.js server and run populartimes.js there. I will update after I give that a try. Thank you!

**Comment by zhg-kj** [comment]
I was able to get it to work by following the steps to create a backend here: https://dev.to/nburgess/creating-a-react-app-with-react-router-and-an-express-backend-33l3

and then implenting populartimes.js there. I am curious, is the now property in your program scraping the data for the live popular time? (the returned percentage is different from what Google is displaying as live)

**Comment by BlakeStevenson** [comment]
I wrote this in an afternoon, just to see if I could get it working, as such I have not done very much testing. It is possible that there is an error in the `getPopularTimes` function, but I was unable to notice any, since I am explicitly checking if the first word of the index in each object is equal to "Currently" using:
```
if(parts[0] === "Currently") {
    out.now = {
        percent: parts[4],
    }
}
```
(parts being Currently something something something x%)
This will only show up on in the return value when Google has live data, which they sometimes don't
In theory, yes, it should be scraping the live data from Google Maps, however, I just ran a couple of tests, on place ID `ChIJqRNf_2qVwoAR8LxgjIId7Qw`, and the `now` index indeed does not match what is available online.

**Comment by BlakeStevenson** [comment]
Ok, so I am actually very dumb, I looked again in inspect element and the `aria-label` of that graph bar contains `Currently 41% busy, usually 81% busy.`, and I was adding what it usually is instead of what it actually is. I've changed the format of the `now` index to be:
`
{
    "currently": "x%",
    "usually": "y%"
}
`
I have fixed this in commit https://github.com/BlakeStevenson/populartimes.js/commit/60552a79141b3f1133f1bdcdfa25db8b8c55fb2c. Thanks for bringing this to my attention.
